lms/djangoapps/enthought/auth.py

```diff
@@ -0,0 +1,169 @@
+""" Authenticate the user using Enthought credentials. """
+
+# Standard library.
+import os
+import requests
+
+# 3rd party library.
+from django import forms
+from django.conf import settings
+from django.contrib.auth.models import User
+
+# edX library.
+from student.forms import AccountCreationForm
+
+
+class EnthoughtAuthBackend(object):
+    """ A Django authentication backend that authenticates users via the
+    Enthought web API.
+
+    """
+
+    def authenticate(self, username=None, password=None, request=None):
+        """
+        Try to authenticate a user. This method will try to authenticate the
+        user on the Enthought API system and if successful, return a User
+        object from the local database.
+
+        If a User object is not found in the local database, it creates one.
+        """
+
+        if not username:
+            username = request.POST.get('email')
+
+        if not password:
+            password = request.POST.get('password')
+
+        status, user_data = self._authenticate(username, password)
+
+        if status == 'invalid':
+            user = None
+
+        elif status == 'inactive':
+            user = User(username=username, is_active=False)
+
+        else:
+            try:
+                user = User.objects.get(username=username)
+
+            except User.DoesNotExist:
+                user = self._create_user(
+                    username   = username,
+                    password   = password,
+                    first_name = user_data['first_name'],
+                    last_name  = user_data['last_name'],
+                    is_active  = user_data['is_active']
+                )
+
+        return user
+
+    def get_user(self, user_id):
+        """ Return the User object for a user that has already been
+        authenticated by this backend.
+
+        """
+
+        try:
+            return User.objects.get(id=user_id)
+
+        except User.DoesNotExist:
+            return None
+
+    #### Private protocol #####################################################
+
+    _API_ROOT   = 'https://api.enthought.com'
+
+    def _authenticate(self, username, password):
+        """ Authenticate the user on api.enthought.com. """
+
+        url  = self._API_ROOT + '/token-auth/'
+        data = {'username': username, 'password': password}
+
+        authenticated = requests.post(url, data=data)
+        user_data     = self._get_user_data(username)
+
+        if authenticated.ok:
+            status = 'success'
+
+        else:
+            if user_data.get('is_active') is False:
+                status = 'inactive'
+
+            else:
+                status = 'invalid'
+
+        return status, user_data
```

Comment:
If `authenticated.ok` then `status, user_data` is returned. But we don't know what the contents of `user_data` is, or if that call even succeeded - but let's say it's an empty dict. A this point `authenticate()` gets a status of `success` and an empty dictionary which means lines 53, 54, 55 will raise `KeyError`s which will result in a 500.

The logic here is broken and it is confusing to read and there is important code missing (like handling responses from API). Debugging this will be hard, and there is no logging anywhere so we wouldn't know why a call to get_user_data failed - or even if it did fail.